src/tags/models.py

```diff
@@ -1,8 +1,29 @@
 from django.db import models
 
+from django.apps import apps
 from taggit.managers import TaggableManager
 
 
+def total_xp_by_tags(user, tags: list):
+    """ 
+        Returns user's total xp over active_semester filtered by tags
+    """ 
+    # get models through here to prevent circular imports
+    Quest = apps.get_model("quest_manager", "Quest")
+    QuestSubmission = apps.get_model("quest_manager", "QuestSubmission")
+    Badge = apps.get_model("badges", "Badge")
+    BadgeAssertion = apps.get_model("badges", "BadgeAssertion")
+
+    # get all objects with xp that is related to user
+    # convert submission/assertion qs into Quest/Badge qs
+    xpitems = [
+        Quest.objects.filter(id__in=[qs.quest.id for qs in QuestSubmission.objects.all_completed(user=user)], tags__name__in=tags).distinct(), 
+        Badge.objects.filter(id__in=[ba.badge.id for ba in BadgeAssertion.objects.all_for_user(user=user)], tags__name__in=tags).distinct(),
```

Comment:
Does this handle cases where a student earns multiple assertions of the same badge, or submits a repeatable quests mutliple times? Do they get all the XP for those?  I didn't see a test for that.

Reply:
I think you might be using the wrong Submission method. "Completed" for submissions means everything that shows up in the completed tab for a student. I.e. quest has been submitted by the student but might be approved by a teacher already, or might be awaiting approval). I think you want all_approved()? Please double check (reviewing from my phone :)

Sorry my method docstrings were very bad back when I wrote that stuff!

Reply:
Also, surprised this wasn't caught in testing.

Reply:
function now accounts for ordinal for badge assertions and quest submission + tests,
changed all_completed to all_approved()